### ISSUE TITLE
[RFC]: Deprecate the FlexAttention backend

### ISSUE THREAD

**Issue by yewentao256** [vllm-project/vllm#50324]
### Motivation.

FlexAttention is no longer selected by default on ROCm and ranks behind TritonAttention on CUDA.

It has a lot of test coverage and is not used in practice, and sometimes breaks CI like https://github.com/vllm-project/vllm/issues/46531

**TritonAttention** now covers the standard FlexAttention use cases, including FP32, encoder-only, non-causal, sliding window, multimodal prefix, batch invariance, and Unlimited-OCR R-SWA.

Removing FlexAttention would eliminate roughly 1,900 lines of backend-specific implementation and tests, while reducing reliance on private PyTorch FlexAttention APIs.

### Proposed Change.

- Now: Mark `FLEX_ATTENTION` as deprecated.
- v0.29.0/v0.30.0: Remove the backend implementation, enum entry, Flex-specific configuration fields, and tests.

### CC List.

@liangel-02 @Isotr0py @MatthewBonanni @mgoin @LucasWilkinson @andakai

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by zou3519** [comment]
cc @drisspg

**Comment by AndreasKaratzas** [comment]
Flex attention had good potential here to be a batch invariant backend. It was cross platform and accurate. I know that it is not widely used anymore so if it is creating CI problems then perhaps we should indeed deprecate it. I haven't captured flex attention problems tbh on CI.
cc @khluu

**Comment by yewentao256** [comment]
Thanks @AndreasKaratzas. Note that for batch invariance it is not a must have as well, we have triton attention or FA as more reliable backends.

**Comment by drisspg** [comment]
🫡

**Comment by mgoin** [comment]
Are there examples of flaky CI other than batch invariance? We can just stop advertising flex attention as supporting that feature. I haven't seen other issues?

**Comment by MatthewBonanni** [comment]
I haven't really run into too many CI issues with it; I'm not necessarily opposed to removing it but I don't feel like it's been much of a maintenance burden

**Comment by yewentao256** [comment]
@mgoin @MatthewBonanni Here are some issues related
- [#27724](https://github.com/vllm-project/vllm/issues/27724)：fixed by [#27853](https://github.com/vllm-project/vllm/pull/27853) 
- [#27344](https://github.com/vllm-project/vllm/pull/27344)：Bumping to PyTorch 2.9 ，Flex  `_build_block_mask_direct` returns the wrong shape
- [#41342](https://github.com/vllm-project/vllm/issues/41342)：AMD pooling nightly with FLEX_ATTENTION case failures
- And of course https://github.com/vllm-project/vllm/issues/46531

**Comment by MatthewBonanni** [comment]
I mean, 2 of those were from last October and #41342 was just an issue with the test itself. Like I said, not opposed to deprecating, I just don't think CI is the reason necessarily

**Comment by yewentao256** [comment]
Thanks @MatthewBonanni ! There are some issues with some feature migration as well like v1 to v2 #39353.
Yes CI is not the only reason, the main thing is that it might not worth the effort to maintain a backend that is seldom use in practice by most of the users. And that's why I raise this RFC, hoping to get more feedbacks from community, eg. some users might say they use it a lot.

**Comment by liangel-02** [comment]
> the main thing is that it might not worth the effort to maintain a backend that is seldom use in practice by most of the users

hey @yewentao256, torchtitan depends on vLLM's flex attention to support the generator side of their RL loop (see example config [here](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/examples/alphabet_sort/config_registry.py#L194)) and it has been very helpful for us to have an implementation that integrates with vLLM out of the box!

**Comment by Isotr0py** [comment]
FlexAttention provides a flexible interface to draft/invesitgate sparse attention mask support. For example, Gemma3 bidi attn and Unlimited-OCR R-SWA support are both firstly implemented on FlexAttention backend, then migrated to Triton/FA.

I think a more critical question is that whether there will be a sparse mask implementation, which is difficult to be implemented on FA/Triton but not for FlexAttn.

**Comment by aoshen02** [comment]
I think it does provide a flexible way to config customized attention, especially some people in RL use it to do some validation. Maybe we should introduce magi attention to replace it if we want to remove this.

**Comment by yewentao256** [comment]
As there are some users now currently using this feature, let's not deprecate it for now, thanks everyone for the discussion!
